File: python_code/src/dynamic_multiplex/fit_multilayer_hungarian.py
from __future__ import annotations

import numpy as np

from .multilayer_utils import (
    fit_layer_communities,
    make_layer_links,
    prepare_multilayer_graphs,
)


def _assign_max_overlap(overlap: np.ndarray) -> np.ndarray:
    """Return, for each row, the column it is matched to so that total overlap
    is maximised (Hungarian / optimal assignment). Falls back to a greedy
    matcher if SciPy is unavailable."""
    d = max(overlap.shape)
    square = np.zeros((d, d), dtype=int)
    square[: overlap.shape[0], : overlap.shape[1]] = overlap
    try:
        from scipy.optimize import linear_sum_assignment

        rows, cols = linear_sum_assignment(square.max() - square)
        col_for_row = np.empty(d, dtype=int)
        col_for_row[rows] = cols
        return col_for_row
    except Exception:  # pragma: no cover - greedy fallback
        used: set[int] = set()
        col_for_row = np.full(d, -1, dtype=int)
        for i in np.argsort(-square.max(axis=1)):
            for c in np.argsort(-square[i]):
                c = int(c)
                if c not in used:
                    col_for_row[i] = c
                    used.add(c)
                    break
        return col_for_row


def _match_communities_hungarian(memberships: list[np.ndarray]) -> list[np.ndarray]:
    """Align per-layer memberships by Hungarian matching of consecutive layers
    on the community overlap (contingency) matrix. Births get fresh labels."""
    out = [np.asarray(memberships[0], dtype=int)]
    next_free = int(out[0].max()) + 1 if out[0].size else 1

    for t in range(1, len(memberships)):
        prev = out[t - 1]
        cur = np.asarray(memberships[t], dtype=int)
        cl = np.unique(cur)
        pl = np.unique(prev)

        overlap = np.zeros((len(cl), len(pl)), dtype=int)
        for i, ci in enumerate(cl):
            for j, pj in enumerate(pl):
                overlap[i, j] = int(np.sum((cur == ci) & (prev == pj)))

        assignment = _assign_max_overlap(overlap)

        relabel: dict[int, int] = {}
        for i, ci in enumerate(cl):
            col = int(assignment[i])
            if col < len(pl) and overlap[i, col] > 0:
                relabel[int(ci)] = int(pl[col])
        for ci in cl:
            if int(ci) not in relabel:
                relabel[int(ci)] = next_free
                next_free += 1
        next_free = max(next_free, max(relabel.values()) + 1)

        out.append(np.array([relabel[int(v)] for v in cur], dtype=int))

    return out


def fit_multilayer_hungarian(
    layers,
    algorithm: str = "leiden",
    resolution_parameter: float = 1.0,
    directed: bool = False,
    objective: str | None = None,
):
    """Fit multilayer communities with Hungarian snapshot matching.

    Detects communities independently in each layer, then tracks them across
    time by matching community labels between consecutive layers with the
    Hungarian (optimal assignment) algorithm on the community overlap matrix.
    Unlike the coupling-based fits (``fit_multilayer_jaccard``,
    ``fit_multilayer_overlap``, ``fit_multilayer_identity_ties``), this is a
    two-stage snapshot-and-match tracker: communities are found per layer and
    aligned post hoc, not jointly optimised on a coupled supra-graph.

    Returns
    -------
    dict
        Keys ``layer_communities`` (per-layer detection), ``meta_communities``
        (one array per layer with labels aligned across consecutive layers by
        Hungarian matching; see ``extract_meta_membership``), ``meta_ids``,
        ``interlayer_ties`` (``None`` -- no supra-graph is built), ``method``
        (``"hungarian"``), and ``layer_links``.
    """
    graph_layers = prepare_multilayer_graphs(layers, directed=directed)
    links = make_layer_links(len(graph_layers), None)

    fit = fit_layer_communities(
        graph_layers,
        algorithm=algorithm,
        resolution_parameter=resolution_parameter,
        directed=directed,
        objective=objective,
    )

    memberships = [
        np.array([f.membership[n] for n in sorted(f.membership)], dtype=int) for f in fit
    ]
    meta = _match_communities_hungarian(memberships)
    meta_ids = sorted({int(v) for layer in meta for v in layer})

    return {
        "algorithm": algorithm,
        "layer_communities": fit,
        "meta_communities": meta,
        "meta_ids": meta_ids,
        "layer_links": links,
        "interlayer_ties": None,
        "method": "hungarian",
        "directed": directed,
    }

File: r_code/R/fit_multilayer_hungarian.R
#' @title Fit multilayer communities with Hungarian snapshot matching
#'
#' @description Detects communities independently in each layer and then tracks
#' them across time by matching community labels between consecutive layers with
#' the Hungarian (optimal linear-sum assignment) algorithm applied to the
#' community overlap (contingency) matrix. This is a two-stage
#' \emph{snapshot-and-match} tracker: unlike the coupling-based fits
#' (\code{\link{fit_multilayer_jaccard}}, \code{\link{fit_multilayer_overlap}},
#' \code{\link{fit_multilayer_identity_ties}}), communities are found per layer
#' and aligned \emph{post hoc}, not jointly optimised on a coupled supra-graph.
#' It provides the common "independent detection + optimal label matching"
#' baseline (e.g. per-layer Leiden matched with the Hungarian algorithm) as a
#' first-class method with the same interface as the other fits.
#'
#' @param layers List of \code{igraph} objects or square adjacency matrices.
#'
#' @param algorithm Community algorithm: \code{"louvain"} or \code{"leiden"}.
#'
#' @param resolution_parameter Leiden resolution parameter.
#'
#' @param directed Logical; if \code{TRUE}, build directed graphs from adjacency
#' matrices (collapsed to undirected for detection, as elsewhere in the
#' package).
#'
#' @param objective One of \code{"cpm"} or \code{"modularity"} (directed
#' networks only).
#'
#' @param seed Optional integer seed for reproducible community detection.
#' When supplied, the global RNG state is saved, the RNG is seeded for the
#' duration of the call, and the previous state is restored on exit, so the
#' caller's random number stream (e.g. bootstrap resampling) is unaffected.
#' Defaults to `NULL` (detection inherits the caller's RNG stream, matching
#' previous behavior; call `set.seed()` beforehand for reproducibility).
#'
#' @return A list of class \code{"multilayer_community_fit"} with components:
#'   \describe{
#'     \item{layer_communities}{Per-layer community detection (each with
#'       \code{membership} and \code{communities}), detected independently.}
#'     \item{meta_communities}{The tracked partition: one integer vector per
#'       layer where labels are aligned across consecutive layers by Hungarian
#'       matching on community overlap. A community with no positive-overlap
#'       match is given a new label (a birth); labels not carried forward are
#'       treated as deaths. See \code{\link{extract_meta_membership}}.}
#'     \item{interlayer_ties}{\code{NULL} - no coupled supra-graph is built.}
#'     \item{method}{\code{"hungarian"}.}
#'     \item{layer_links}{Sequential adjacent-layer links (the chain the
#'       matcher walks).}
#'   }
#'
#' @details Matching is sequential: layer \eqn{t} is aligned to layer
#' \eqn{t-1}. For each adjacent pair the community overlap matrix is completed
#' to a square cost matrix and \code{\link[clue]{solve_LSAP}} maximises total
#' overlap; unmatched current communities receive fresh labels. The result is a
#' consistently-labelled sequence suitable for \code{\link{extract_meta_membership}},
#' \code{\link{bootstrap_multilayer}}, and the plotting helpers.
#'
#' @seealso \code{\link{fit_multilayer_jaccard}},
#'   \code{\link{fit_multilayer_overlap}},
#'   \code{\link{fit_multilayer_identity_ties}} for coupling-based tracking.
#'
#' @examples
#' set.seed(123)
#' layers <- lapply(1:3, function(i) {
#'   m <- matrix(rbinom(64, 1, 0.35), nrow = 8)
#'   m <- pmax(m, t(m)); diag(m) <- 0; m
#' })
#' fit <- fit_multilayer_hungarian(layers, algorithm = "leiden")
#' extract_meta_membership(fit)
#'
#' @export

fit_multilayer_hungarian <- function(
    layers,
    algorithm = c("louvain", "leiden"),
    resolution_parameter = 1,
    directed = FALSE,
    objective = NULL,
    seed = NULL
  ) {

  algorithm <- match.arg(algorithm)

  # Scoped seed for reproducible detection (restores caller RNG state) ----
  if (!is.null(seed)) {
    rng_state <- save_rng_state()
    on.exit(restore_rng_state(rng_state), add = TRUE)
    set.seed(seed)
  }

  # Prepare graph layers ----
  graph_layers <- prepare_multilayer_graphs(layers, directed = directed)
  links <- make_layer_links(length(graph_layers), NULL)

  # Stage 1: independent per-layer community detection ----
  fit <- fit_layer_communities(
    graph_layers,
    algorithm = algorithm,
    resolution_parameter = resolution_parameter,
    directed = directed,
    objective = objective
  )

  # Stage 2: Hungarian label matching across consecutive layers ----
  memberships <- lapply(fit, function(x) as.integer(x$membership))
  meta <- match_communities_hungarian(memberships)

  # Compile multilayer community fit object ----
  multilayer_hungarian <- structure(
    list(
      algorithm = algorithm,
      layer_communities = fit,
      meta_communities = meta,
      meta_ids = sort(unique(unlist(meta))),
      layer_links = links,
      interlayer_ties = NULL,
      method = "hungarian",
      directed = directed
    ),
    class = "multilayer_community_fit"
  )

  return(multilayer_hungarian)
}


# Internal: align a sequence of per-layer memberships by Hungarian matching of
# consecutive layers on the community overlap matrix (maximise total overlap).
match_communities_hungarian <- function(memberships) {

  out <- vector("list", length(memberships))
  out[[1L]] <- as.integer(memberships[[1L]])
  next_free <- max(c(out[[1L]], 0L)) + 1L

  if (length(memberships) >= 2L) {
    for (t in 2:length(memberships)) {

      prev <- out[[t - 1L]]
      cur  <- as.integer(memberships[[t]])
      cl   <- sort(unique(cur))
      pl   <- sort(unique(prev))

      # community overlap (contingency) matrix ----
      overlap <- matrix(0L, nrow = length(cl), ncol = length(pl))
      for (i in seq_along(cl)) {
        for (j in seq_along(pl)) {
          overlap[i, j] <- sum(cur == cl[i] & prev == pl[j])
        }
      }

      # pad to a square cost matrix and solve the assignment ----
      d <- max(nrow(overlap), ncol(overlap))
      square <- matrix(0L, d, d)
      square[seq_len(nrow(overlap)), seq_len(ncol(overlap))] <- overlap
      assignment <- as.integer(clue::solve_LSAP(max(square) - square))

      # map current labels to previous labels; births get fresh labels ----
      relabel <- rep(NA_integer_, length(cl))
      for (i in seq_along(cl)) {
        col <- assignment[i]
        if (col <= length(pl) && overlap[i, col] > 0L) relabel[i] <- pl[col]
      }
      for (i in seq_along(cl)) {
        if (is.na(relabel[i])) {
          relabel[i]  <- next_free
          next_free   <- next_free + 1L
        }
      }
      next_free <- max(next_free, max(relabel) + 1L)

      names(relabel) <- as.character(cl)
      out[[t]] <- unname(relabel[as.character(cur)])
    }
  }

  lapply(out, as.integer)
}
